drafts/refactoring-legacy-code-with-tiny-types-in-java.md

```diff
@@ -0,0 +1,237 @@
+
+  public static UnaryOperator<BoundedQuality> adjustQuality(int adjustment) {
+    return it -> new BoundedQuality(Math.max(0, Math.min(50, it.value + adjustment)));
+  }
+}
+```
+
+Whenever a method receives a `BoundedQuality`, we can confidently assert that the data's integrity remains intact, allowing us to remove unnecessary defensive checks.
+Additionally, the `adjust` method enables safe manipulation of the quality amount.
+Moreover, if we attempt to create a `BoundedQuality` with invalid data, such as a negative amount, the system fails fast.
+
+Before we introduced the `BoundedQuality` type, we modeled both the `quality` and `sellIn` amounts as primitive integers.
+The compiler couldn’t help us distinguish between these two numbers since they share the same type.
+This situation could lead to bugs if we passed them out of order to a function like `doSomething(int, int)` by accident.
+
+But hang on! What do we do with this new requirement?
+
+* *"Sulfuras"* is a legendary item and as such its `Quality` is `80` and it never alters
+* *"Sulfuras"* is a legendary item, never has to be sold or decreases in `Quality`
+
+We can introduce yet another tiny type, how about a `FixedQuality`? Most likely we want to use the qualities interchangeably and introduce an interface as well.
+
+```java
+public sealed interface Quality permits BoundedQuality, FixedQuality {
+  int value();
+}
+
+record BoundedQuality(int value) implements Quality { /* ... */ }
+
+record FixedQuality(int value) implements Quality { 
+  public static final FixedQuality EIGHTY = new FixedQuality(80);
+}
+```
+
+As you can see, we use a `sealed` interface introduced with Java 17.
+[Sealed interfaces](https://openjdk.org/jeps/409) complement Java's enhanced [pattern matching](https://openjdk.org/jeps/441) by providing exhaustive case handling that the compiler enforces.
+This discussion naturally leads us to another topic I'd like to cover: [Data Oriented Programming in Java](https://www.infoq.com/articles/data-oriented-programming-java/), as explained in an InfoQ article by Brian Goetz.
+
+> _Data-oriented programming_ encourages us to model data as (immutable) data, and keep the code that embodies the business logic of how we act on that data separately.
+Records, sealed classes, and pattern matching are designed to work together to support data-oriented programming.
+Using the combination of records and sealed classes also makes it easier to _make illegal states unrepresentable_, further improving safety and maintainability.
+
+In the spirit of making illegal states unrepresentable, modeling our purchasable items from the shop begins here.
+For example, *"Sulfuras"* has a quality of `80` that never changes.
+We model our legendary item as a Java Record because these records are immutable by default, allowing us to hard-code the quality.
+
+Sealed interfaces help with the creation of a lightweight type hierarchy, allowing us to group all the items together for processing.
```

Comment:
IMO it would read better to explicitly refer to PurchasableItem, ie: 
_allowing us to group all the items together for processing as 'PurchasableItem'_